### PR TITLE
Grab tag graph 47 - Not quite clean. In progress.

### PR DESCRIPTION
**Pull request by meldefon**

### CODE REVIEW

**Comment by meldefon** [comment]
@kwhitehall  In principle, I agree with your comment about moving the CeFinder class into utils. I just ran into problems with some circular including. It may be easier to leave the class in mccSearch. If that's a huge problem, I'll find a workaround. Just let me know.

**Comment by meldefon** [comment]
@kwhitehall  I made the cosmetic changes you asked for (except moving the class, which is discussed above). Next priority is getting you the flowchart you asked for. I'll keep you updated.

**Comment by kwhitehall** [comment]
Hi @meldefon nice work. There is however still a significant bug to be addressed, which is multiple entries for cloud element data in cloudElement.txt and cloudElementsUserFile.txt are being made. Only one entry should be in these files per a cloud element.

File: code/mccSearch.py [hunk 1]
```diff
@@ -542,7 +612,838 @@ def find_cloud_elements(mergImgs, timelist, mainStrDir, lat, lon, TRMMdirName=No
...
+            #tmax = ndimage.maximum(cloudElement, labels=labels)*1.
+            #if edges:
+            #    cloudElementsJSON.append({'cloudElement': ceUniqueID, 'time': str(timelist[t]),\
+            #        'area':cloudElementArea, 'Tmax': tmax, 'Tmin': tmin,'center':cloudElementCenter,\
+            #        'convective_fraction': cf, 'lat_lon_box': latLonBox, 'shape': shape, 'edges':edges, 'eccentricity':cloudElementEpsilon})
+
+            profTimes[8]+= time.time() - pin
+            pin = time.time() #tenth pin drop
+
+            if frameNum==1:
```
Comment (kwhitehall) [inline code/mccSearch.py:1359]: This if statement is incorrect. I believe you mis-indented from the original code. According to the original code, this else statement was not for the "if frameNum != 1", but for the if statement "if cloudElementArea >= AREA_MIN or (cloudElementArea < AREA_MIN and \
                    ((ndimage.minimum(cloudElement, labels=labels)) / \
                        float((ndimage.maximum(cloudElement, labels=labels)))) < CONVECTIVE_FRACTION):"

File: code/mccSearch.py [hunk 2]
```diff
@@ -542,7 +612,838 @@ def find_cloud_elements(mergImgs, timelist, mainStrDir, lat, lon, TRMMdirName=No
...
+                    'cloudElementTmax':TIR_max, 'cloudElementTmin': TIR_min,}
+   
+            #Store the new cloudElementDict so that it can be returned to parent function
+            allCloudElementDicts.append(cloudElementDict)
+            
+            #Will hold data for JSON
+            currentJSONDict = {}
+            
+            # get some data for the JSON object which will only store the graph CEs and connected edges
+            currentJSONDict['cf'] = (((ndimage.minimum(cloudElement, \
```
Comment (kwhitehall) [inline code/mccSearch.py:1334]: @meldefon the logic associated with creating this JSON object is incorrect

File: code/mccSearch.py [hunk 3]
```diff
@@ -542,7 +612,838 @@ def find_cloud_elements(mergImgs, timelist, mainStrDir, lat, lon, TRMMdirName=No
...
+
+
+            #This replaces the loop computation of brightnesstemp1, commented lines are a test
+            brightnesstemp1[0,loc[0],loc[1]] = cloudElement
+            #Mini unit test comparing this to loop version. To use, uncomment the loop computation
+            #assert(np.array_equal(brightnesstemp1,brightnesstemp1Old))
+
+            
+            #This replaces the loop computation of finalCETRMMvalues and ceTRMM, tests commented
+            if TRMMdirName:
```
Comment (kwhitehall) [inline code/mccSearch.py:1191]: These "if TRMMdirName:" statements should be as best as possible made into one.

File: code/mccSearch.py [hunk 4]
```diff
@@ -542,7 +612,838 @@ def find_cloud_elements(mergImgs, timelist, mainStrDir, lat, lon, TRMMdirName=No
...
+        profTimes[1]+= time.time() - pin
+        
+
+        #If the area is greater than the area required, or if the area is smaller than the suggested area,
+        #check if it meets a convective fraction requirement consider as CE
+
+        if cloudElementArea >= AREA_MIN or (cloudElementArea < AREA_MIN and \
+                ((ndimage.minimum(cloudElement, labels=labels)) / \
+                    float((ndimage.maximum(cloudElement, labels=labels)))) < CONVECTIVE_FRACTION):
+            pin = time.time() #third pin drop
```
Comment (kwhitehall) [inline code/mccSearch.py:982]: @meldefon Why are these pins here. Is this part of your testing?

File: code/mccSearch.py [hunk 5]
```diff
@@ -542,7 +612,838 @@ def find_cloud_elements(mergImgs, timelist, mainStrDir, lat, lon, TRMMdirName=No
...
+    # cloudElementsTextFile = open(thisFileName,'w')
+    #-------------------------------------------------
+
+    #determine contiguous locations with temeperature below the warmest temp i.e. cloudElements in each frame
+    frame, ceCounter = ndimage.measurements.label(mergImgs[t,:,:], structure=STRUCTURING_ELEMENT)
+    #frame, ceCounter = ndimage.measurements.label(Img, structure=STRUCTURING_ELEMENT)
+    frameceCounter = 0
+    frameNum = t + 1
+
+    profTimes = np.zeros((10,))
```
Comment (kwhitehall) [inline code/mccSearch.py:880]: Again, is this variable and the pins for?

File: code/mccSearch.py [hunk 6]
```diff
@@ -542,7 +612,838 @@ def find_cloud_elements(mergImgs, timelist, mainStrDir, lat, lon, TRMMdirName=No
...
+            nonEmptyLons = sum(sum(cloudElement) > 0)
+            nonEmptyLats = sum(sum(cloudElement.transpose()) > 0)
+
+            shape = max(nonEmptyLats, nonEmptyLons)
+            
+            cloudElementsUserFileString+=('\nCenter (lat,lon) is: %.2f\t%.2f' %(latCenter, lonCenter))
+            cloudElementCenter.append(latCenter)
+            cloudElementCenter.append(lonCenter)
+            cloudElementsUserFileString+=('\nNumber of boxes are: %d' %numOfBoxes)
+            cloudElementsUserFileString+=('\nArea is: %.4f km^2' %(cloudElementArea))
```
Comment (kwhitehall) [inline code/mccSearch.py:1301]: @cgoodale do you think this is resource efficient to create a long string with the data to write in the file?

Comment (cgoodale) [inline code/mccSearch.py:1301]: Hey @kwhitehall and @meldefon the best approach would be to define the string once all the input variable values are known.  You can easily build a multi-line string and call the .format() built-in function.

Here is a good cookbook I found online:
https://mkaz.github.io/2012/10/10/python-string-format/

The python docs on this are a bit too cryptic and dense for me.

Example:  
"""Number of boxes are: {boxCount:d}
Area is: {cloudArea:.4f} km^2""".format(boxCount= numOfBoxes, cloudArea= cloudElementArea)

**Pull request merged**